File: python/fracgm/registration.py
import numpy as np
import scipy as sp

from fracgm.utils import compute_initial_guess, compute_terms, project
from fracgm.utils import se3_mat_to_vec, se3_vec_to_mat


class FracGM:
    def __init__(self, max_iteration=1000, tol=1e-7, c=1, noise_bound=0.1):
        self.max_iter = max_iteration
        self.tol = tol
        self.c = c
        self.noise_bound = noise_bound

    def solve_beta_mu(self, terms):
        beta = [term.f() / term.h() for term in terms]
        mu = [1 / term.h() for term in terms]

        return beta, mu

    def solve_x(self, beta, mu, terms):
        mat_a = np.zeros((13, 13))
        for i in range(self.num):
            mat_a += mu[i] * terms[i].f_mat() - mu[i] * beta[i] * terms[i].h_mat()

        lu, piv = sp.linalg.lu_factor(mat_a)
        temp = sp.linalg.lu_solve((lu, piv), self.e)
        return (1 / (self.e @ temp)) * temp

    def compute_psi_norm(self, beta, mu, terms):
        assert len(beta) == len(mu)
        assert len(beta) == len(terms)

        loss = 0.0
        for i, term in enumerate(terms):
            a = -term.f() + beta[i] * term.h()
            b = -1.0 + mu[i] * term.h()
            loss += a * a + b * b

        return np.sqrt(loss)

    def update_terms_cache(self, terms, vec):
        for term in terms:
            term.update_cache(vec)

    def solve(self, pcd1, pcd2):
        assert pcd1.shape == pcd2.shape, "Input point clouds must have the same shape."
        self.num = pcd1.shape[0]
        self.e = np.array(
            [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0]
        )

        terms = compute_terms(pcd1, pcd2, self.noise_bound**2, self.c**2)

        init_mat = compute_initial_guess(pcd1, pcd2)
        x = se3_mat_to_vec(init_mat)
        beta, mu = self.solve_beta_mu(terms)

        for _ in range(self.max_iter):
            # alternating solve x
            x = self.solve_x(beta, mu, terms)
            self.update_terms_cache(terms, x)

            # stopping criteria
            psi_norm = self.compute_psi_norm(beta, mu, terms)
            if psi_norm < self.tol:
                break

            # alternating solve beta & mu
            beta, mu = self.solve_beta_mu(terms)

        se3 = se3_vec_to_mat(x)
        se3[:3, :3] = project(se3[:3, :3])

        return se3

File: cpp/fracgm/src/registration.cpp
#include <Eigen/Dense>
#include <cmath>
#include <vector>

#include "fracgm/constant.h"
#include "fracgm/registration.h"
#include "fracgm/utils.h"

namespace fracgm {

registration::registration(const int &max_iteration, const double &tol,
                           const double &c, const double &noise_bound) {

  max_iteration_ = max_iteration;
  tol_ = tol;
  c_ = c;
  noise_bound_ = noise_bound;
}

std::vector<double>
registration::solve_beta_mu(std::vector<Fractional> *terms) {
  std::vector<double> vec_1; // beta
  std::vector<double> vec_2; // mu

  for (auto &term : *terms) {
    vec_1.push_back(term.f() / term.h());
    vec_2.push_back(1 / term.h());
  }

  vec_1.insert(vec_1.end(), vec_2.begin(), vec_2.end());
  return vec_1;
}

Eigen::VectorXd registration::solve_x(std::vector<double> *alpha,
                                      std::vector<Fractional> *terms) {

  Eigen::MatrixXd mat_a = Eigen::MatrixXd::Zero(13, 13);
  int n = terms->size();

  for (int i = 0; i < n; i++) {
    mat_a += alpha->at(n + i) * terms->at(i).f_mat() -
             alpha->at(n + i) * alpha->at(i) * terms->at(i).h_mat();
  }

  Eigen::Vector<double, 13> temp = mat_a.ldlt().solve(e);
  return (1 / (e.transpose() * temp)) * temp;
}

float registration::compute_psi_norm(std::vector<double> *alpha,
                                     std::vector<Fractional> *terms) {
  float a;
  float b;
  int n = terms->size();

  float loss = 0.0;
  for (int i = 0; i < n; i++) {
    a = -terms->at(i).f() + alpha->at(i) * terms->at(i).h();
    b = -1.0 + alpha->at(n + i) * terms->at(i).h();
    loss += a * a + b * b;
  }
  return sqrt(loss);
}

void registration::update_terms_cache(std::vector<Fractional> *terms,
                                      Eigen::VectorXd *vec) {
  for (auto &term : *terms) {
    term.update_cache(*vec);
  }
}

Eigen::Matrix4d registration::solve(const PointCloud &pcd1,
                                    const PointCloud &pcd2) {

  Eigen::VectorXd x;
  Eigen::Matrix4d se3;
  std::vector<double> alpha;
  float psi_norm;

  std::vector<Fractional> terms =
      compute_terms(pcd1, pcd2, noise_bound_ * noise_bound_, c_ * c_);

  Eigen::Matrix4d init_mat = compute_initial_guess(pcd1, pcd2);
  x = se3_mat_to_vec(init_mat);

  for (int i = 0; i < max_iteration_; i++) {
    // alternating solve alpha
    alpha = solve_beta_mu(&terms);
    // alternating solve x
    x = solve_x(&alpha, &terms);
    update_terms_cache(&terms, &x);
    // stopping criteria
    psi_norm = compute_psi_norm(&alpha, &terms);
    if (psi_norm < tol_) {
      break;
    }
  }

  se3 = se3_vec_to_mat(x);
  se3.block<3, 3>(0, 0) = project(se3.block<3, 3>(0, 0));
  return se3;
}

} // namespace fracgm